### PR TITLE
Normalize opt-in multi-valued JWT claims for Cedar

### LINKED ISSUE #5828: feat(authz/cedar): normalize multi-valued claims to canonical space-delimited form
## Problem

The Cedar authorizer maps a JWT claim to a Cedar value by JSON type: a string becomes a Cedar `String`, an array becomes a `Set`. But the same logical *multi-valued* claim arrives in different shapes across identity providers:

- **Microsoft Entra** emits the scope claim `scp` as one space-delimited string: `"User.Read.All Mail.Read"` (per RFC 6749 §3.3).
- **Keycloak** emits `scope` the same way (space-delimited string).
- **Okta** emits `scp` as a JSON array: `["User.Read.All","Mail.Read"]`.

A Cedar policy that tests element membership can't be written once against both shapes: a `like`-pattern predicate needs a `String`, `.contains()` needs a `Set`, and there is no total Cedar expression across the two (a type mismatch errors, and `||` doesn't swallow a left-operand error). Downstream, this means a policy authored for one IdP's claim shape silently fails closed for another's.

## Change

Add an opt-in list of claim **names** to normalize — `multi_valued_claims` on the `cedarv1` authz config, and `multiValuedClaims` on the vMCP config. The field takes claim names supplied by the caller — not tied to any IdP, so `["scp"]` (Entra/Okta), `["scope"]` (Keycloak), or a custom name all work.

For each listed claim, the authorizer exposes **two** forms, both on the principal entity and the request context:

1. **`claim_<name>` — canonical space-delimited string** (the value's normal `claim_`-prefixed slot):
   - array → join elements with a single space into a `String`;
   - already a `String` → passthrough unchanged;
   - absent → stays absent; not listed → untouched (today's behavior).

2. **`claimset_<name>` — a companion Cedar `Set`** of the claim's elements (array as-is; string split on spaces), for exact-element membership via `.contains()` / `.containsAll()` / `.containsAny()`. The `claimset_` prefix is reserved: because every JWT claim is emitted under `claim_`, a token can never shadow or spoof a `claimset_` attribute.

### Why both a joined string and a companion Set

- **The joined string** lets existing `like`-pattern policies match uniformly across IdPs **without changing the policy text**. That matters where the policy producer and the authorizer runtime upgrade on independent schedules: the emitted predicate stays invariant, so no policy that works today starts failing closed when only one side upgrades. RFC 6749 §3.3 defines scope values as space-free by grammar (`NQCHAR` excludes `%x20`), so joining scope claims is lossless and unambiguous. This is the form the operator-compiled `Contains` predicate (space-delimited element `like`) consumes.
- **The companion Set** is the better representation for *new* multi-value policies: `.contains()` is exact-element (no substring bleed, e.g. `Mail.Read` never matches `Mail.ReadWrite`), `.containsAll()` / `.containsAny()` express multi-scope checks in a single call, and it carries no `like`-metacharacter escaping obligation. It is **additive** — the joined string is unchanged, so nothing that relies on it is affected.

> Note: both forms assume space-**free**-element claims (OAuth scopes). They are **not** appropriate for claims whose elements can contain spaces (e.g. group display names like `"Engineering Team"`), where element boundaries become ambiguous — list only space-free-element claims here. Group/role claims continue to use the entity-hierarchy path, not this field.

## Backward compatibility

Default empty → byte-identical output to today for every existing user, including anyone matching `scp == "a b c"` as a whole string. Purely additive: opt-in per claim name, and even when enabled the original `claim_<name>` string form is preserved (the `Set` is a new attribute, not a replacement).

## Acceptance criteria

- `multi_valued_claims: ["scp"]` + array `scp` → `claim_scp` is a single space-delimited `String` (a `like`-pattern element match succeeds) **and** `claimset_scp` is a `Set` of the elements.
- Same config + string `scp` → `claim_scp` unchanged (passthrough); `claimset_scp` is the space-split `Set`.
- `claimset_scp.contains("X")` matches identically for the array and string shapes, is exact-element (no substring bleed), and works on both `principal.claimset_scp` and `context.claimset_scp`; `claimset_scp` is a bare key, never `claim_claimset_scp`.
- Claim absent → `claim_scp` still absent, no `claimset_scp` fabricated; `has`-guards still evaluate correctly.
- Empty/unset config → byte-identical to current output, no `claimset_` keys (regression test).
- Field plumbed through the `cedarv1` config and the vMCP config.
- Unit tests in `core_test.go`: array-join, string-passthrough, single-element, absent claim, multi-space edge cases, `claimset_` `Set` construction from both shapes, and the bare-key / no-`claim_claimset_` assertion.

## Out of scope / follow-up

Declarative exposure on the operator `AuthzConfigRef` CRD is intentionally deferred (YAGNI): the enterprise compiler writes `multi_valued_claims` straight into the generated Cedar config, and a native operator user can set it today via a raw `MCPAuthzConfig` configMap. A typed CRD field (mirroring `groupClaimName`/`roleClaimName`/`groupEntityType`) can be added later, additively, if a native operator consumer needs it.

### PR DESCRIPTION
**Pull request by jhrozek**
## Summary

OAuth scope claims arrive in different shapes across identity providers: **Entra** (`scp`) and **Keycloak** (`scope`) emit a single space-delimited string, while **Okta** (`scp`) emits a JSON array. The Cedar authorizer maps a string to a Cedar `String` and an array to a `Set`, and there is no single Cedar expression that tests element membership across both shapes — a type mismatch *errors*, and an erroring policy is silently ignored (a `permit` fails closed). So a policy authored for one IdP's claim shape silently mis-authorizes another's.

This adds an **opt-in** `multi_valued_claims` field to the `cedarv1` authz config. For each listed claim name, the authorizer exposes two forms, on both the principal entity and the request context:

- **`claim_<name>`** — a canonical **unpadded space-delimited string** (arrays joined with single spaces; strings passed through unchanged), so existing `like`/`==` string-membership policies match uniformly across IdPs.
- **`claimset_<name>`** — a companion Cedar **`Set`** of the claim's elements, for exact-element membership via `.contains()` / `.containsAll()` / `.containsAny()` (no substring bleed — `Mail.Read` never matches `Mail.ReadWrite`, and no `like`-metacharacter escaping concern).

The `claimset_` prefix is reserved and cannot be shadowed or spoofed by a JWT claim, because every JWT claim is emitted under the `claim_` prefix. The field is opt-in: with it empty, output is **byte-identical** to today, so existing deployments are unaffected.

Closes #5828

## Type of change

- [ ] Bug fix
- [x] New feature
- [ ] Refactoring (no behavior change)
- [ ] Dependency update
- [ ] Documentation
- [ ] Other (describe):

## Test plan

- [x] Unit tests (`task test`)
- [ ] E2E tests (`task test-e2e`)
- [ ] Linting (`task lint-fix`)
- [ ] Manual testing (describe below)

Unit tests were run scoped to the changed package (`go test ./pkg/authz/authorizers/cedar/...`) and are green, covering: array→join, verbatim string passthrough, single-element, empty array, absent claim, non-listed claim, non-string elements, multi-space input; `claimset_` Set construction from both the array and string shapes; exact-element `.contains` with no substring bleed and `.containsAll`; the bare-key / never-`claim_claimset_` assertion; and an empty-config backward-compat regression. Full `task test` and `task lint-fix` run in CI.

## Does this introduce a user-facing change?

Yes. A new optional `multi_valued_claims` list on the `cedarv1` authorization config. When set, each named claim is additionally exposed as a `claimset_<name>` Cedar `Set` and its `claim_<name>` string form is normalized (arrays joined). Unset (default) changes nothing.

## Special notes for reviewers

- **Scope is deliberately the Cedar authorizer only** — no operator CRD field and no vMCP config plumbing. The `claimset_`/join behavior is consumed via the raw `cedarv1` config (the `thv` CLI, a raw `MCPAuthzConfig` configMap, or the enterprise policy compiler, which writes the cedar config directly). A typed operator CRD field was intentionally omitted (YAGNI) — and because `VirtualMCPServerSpec` embeds the vmcp config struct, a vMCP config field would surface in the CRD schema, which we're deliberately avoiding here.
- **Why keep the joined string *and* add a Set:** the joined string preserves the already-shipped `like`-based enterprise `Contains` predicate byte-for-byte, so array-claim (Okta) support lights up as a runtime-only, zero-coordination change with no regression to existing string-IdP (Entra) deployments. The `Set` is the strictly-better representation for *new* multi-value policies and is purely additive.
- Related enterprise work depends on this: stacklok/toolhive#5828 is consumed by the enterprise `Contains` claim operator.

<sub>Generated with [Claude Code](https://claude.com/claude-code)</sub>

### CODE REVIEW

File: pkg/authz/authorizers/cedar/core.go [hunk 1]
```diff
@@ -593,6 +627,113 @@ func preprocessClaims(claims jwt.MapClaims) map[string]interface{} {
...
+		v, ok := claims[name]
+		if !ok {
+			continue
+		}
+
+		switch val := v.(type) {
+		case string:
+			out[name] = val
+		case []interface{}:
+			out[name] = strings.Join(multiValuedTokens(name, val), " ")
```
Comment (JAORMX) [inline pkg/authz/authorizers/cedar/core.go:699]: **[Non-blocking migration hazard — worth a doc warning] Opting an array-shaped claim in flips its `claim_<name>` form from a Cedar `Set` to this joined `String`.** Security + OAuth review both landed on the same consequence: a pre-existing policy like `permit when context.claim_scp like "*Mail.Read*"` that previously *errored* on an array value (Cedar fails closed → deny, per `core.go:509-512`) will, after opt-in, evaluate against the joined string and **substring-match `Mail.ReadWrite` → grant** (fail-open bleed).

Related: the `claim_<name>` **string** input path is a passthrough — the `case string` branch just above re-assigns the already-copied value (a no-op), so the doc's "canonical unpadded" promise only holds for the *array* path; a string IdP's padding/order is preserved verbatim.

Suggest: (a) a migration note — "audit any existing `like` policies on a claim before listing it in `multi_valued_claims`"; (b) steer operators to `claimset_<name>` + `.contains`/`.containsAll`/`.containsAny` for reliable exact-element membership; optionally (c) canonicalize the string path too (`strings.Join(strings.Fields(val), " ")`) so `claim_<name>` is genuinely canonical and the no-op branch does real work.

**Review by JAORMX — DISMISSED** [review]
LGTM — clean, well-scoped opt-in feature, and the security-critical parts hold up. ✅

Reviewed by a three-lens panel (security · OAuth/OIDC · Go/duplication/reuse) against `3067776`, verified against the real code:

**Security (no blockers):** the `claimset_` prefix is genuinely **unspoofable** — `preprocessClaims` prefixes *every* JWT claim with `claim_` (`core.go:624`), and the only writer of bare `claimset_*` keys is operator-config-driven `addMultiValuedClaimSets` reading trusted resolved-token values, so an attacker claim named `claimset_scp` lands inertly at `claim_claimset_scp`. Cedar also **fails closed** on any policy error (`core.go:509-512` returns `false` when `diagnostic.Errors` is non-empty), so type confusion / absent / nested / numeric claims all *deny*. Principal and context surfaces share one `processedClaims` map, so they can't diverge.

**OAuth:** the `claimset_<name>` Set + `.contains`/`.containsAll` model matches RFC 6749 §3.3 scope semantics exactly (order-independent, exact-element — the `Mail.Read` vs `Mail.ReadWrite` test proves no substring bleed). Shape convergence across Entra/Keycloak (`scp`/`scope` string) and Okta (`scp` array) is solid, and the explicit warning against `groups`/`roles` (elements may contain spaces) is the right RFC-vs-reality boundary.

**Go/reuse:** genuine extension, not a parallel path — it reuses the existing `convertToCedarValueAtDepth`/`convertStringArrayToCedar` pipeline for Set construction, respects copy-before-mutate, and is consistent with the #5713 nested-map claim work (nested elements deliberately skipped for flat scope claims). Tests are thorough and table-driven, including the empty-config byte-identical backward-compat assertion.

**Non-blocking follow-ups** (see inline on `core.go:674`):
- **Migration warning:** opting an *array*-shaped claim in flips its `claim_<name>` from Set→String, which can turn a previously-erroring (fail-closed) `like` policy into a fail-open substring match. Worth a CHANGELOG/doc note to audit existing `like` policies before enabling, and steering operators to `claimset_`.
- **Doc vs behavior:** `claim_<name>` is documented "canonical unpadded" but the string-input path is a verbatim passthrough (the `case string` branch is a no-op re-assign) — either canonicalize it (`strings.Join(strings.Fields(val), " ")`) or soften the doc.
- **Harden the invariant in code:** extract `claim_`/`claimset_` into shared consts (currently bare literals at `:624`, `:706-710`) so the collision-freedom relationship is enforced, not incidental.
- Minor: add a note that the `extractGroups` (`:1118`) vs `claimset` type-switches deliberately differ, so nobody "dedups" them; add a string→claimset multi-space test case.

None of these block — approving. CI still running (no failures); merge remains gated on it.

<sub>🤖 AI-assisted panel review via Claude Code (security · OAuth · Go/reuse). Line numbers against `3067776`.</sub>

File: pkg/authz/authorizers/cedar/core.go [hunk 2]
```diff
@@ -593,6 +627,113 @@ func preprocessClaims(claims jwt.MapClaims) map[string]interface{} {
...
+// slog.Debug.
+func multiValuedTokens(claimName string, elems []interface{}) []string {
+	tokens := make([]string, 0, len(elems))
+	for _, elem := range elems {
+		switch elem.(type) {
+		case []interface{}, map[string]interface{}:
+			slog.Debug("multi-valued claim element has unsupported nested type, skipping",
+				"claim", claimName, "type", fmt.Sprintf("%T", elem))
+		default:
+			tokens = append(tokens, fmt.Sprint(elem))
```
Comment (rdimitrov) [inline pkg/authz/authorizers/cedar/core.go]: **Non-string array elements produce spurious tokens, breaking cross-shape parity.**

This `default` branch stringifies any non-`[]`/`map` element. A JSON `null` element yields `fmt.Sprint(nil)` → the literal `"<nil>"`; a numeric element (JSON numbers unmarshal to `float64`) renders via `%g`, so `1000000` → `"1e+06"`.

So `scp: ["Mail.Read", null]` produces `claim_scp = "Mail.Read <nil>"` and `claimset_scp = ["Mail.Read", "<nil>"]`. A policy `context.claim_scp == "Mail.Read"` would then **deny a legitimately-scoped user**, and the array shape diverges from the string shape (whose `strings.Fields` path drops empty tokens) — defeating the feature's stated "works for both IdP shapes" guarantee.

Scopes are space-free strings per RFC 6749, so this is malformed input, but nothing guards it. Consider skipping non-string elements (or empty stringifications) rather than emitting them as tokens.

File: pkg/authz/authorizers/cedar/core.go [hunk 3]
```diff
@@ -593,6 +627,113 @@ func preprocessClaims(claims jwt.MapClaims) map[string]interface{} {
...
+
+	for _, name := range names {
+		v, ok := claims[name]
+		if !ok {
+			continue
+		}
+
+		switch val := v.(type) {
+		case string:
+			out[name] = val
```
Comment (rdimitrov) [inline pkg/authz/authorizers/cedar/core.go:697]: **Redundant no-op.** `out` is already a shallow copy of `claims` (the copy loop above), so `out[name]` already equals `val` here — this assignment is a no-op. The whole `case string` branch can be dropped, letting the string fall through untouched. Minor, but it reads as if the string case does work when it doesn't.

File: pkg/authz/authorizers/cedar/core.go [hunk 4]
```diff
@@ -593,6 +627,113 @@ func preprocessClaims(claims jwt.MapClaims) map[string]interface{} {
...
+// []string (converted to a Cedar Set by convertToCedarValueAtDepth). This is
+// the companion to normalizeMultiValuedClaims's "claim_<name>" string form:
+// where "claim_<name>" supports substring-style `like` matching, "claimset_<name>"
+// supports exact-element `.contains`/`.containsAll`/`.containsAny` matching.
+// The "claimset_" prefix is reserved and cannot collide with a JWT claim, since all
+// JWT claims are "claim_"-prefixed by preprocessClaims.
+//
+// processed must already have the "claim_" prefix applied (i.e. this must run
+// after preprocessClaims) so the added key stays bare. raw is not mutated.
+func addMultiValuedClaimSets(processed map[string]interface{}, raw jwt.MapClaims, names []string) {
```
Comment (rdimitrov) [inline pkg/authz/authorizers/cedar/core.go:722]: **Duplicated type switch / double stringification.** This function and `normalizeMultiValuedClaims` switch on the same raw claim type, and both call `multiValuedTokens` for the `[]interface{}` case — so array elements are stringified twice per authorization call. More importantly, the two parallel 3-arm switches must be kept in sync by hand: a future type case added to one but not the other would silently diverge the `claim_<name>` string form from the `claimset_<name>` Set form. Consider computing the token slice once and deriving both surfaces from it.

**Review by rdimitrov — COMMENTED** [review]
Automated review (medium effort, precision-focused). The design is sound and the test coverage is genuinely thorough — I confirmed the `claimset_` spoofing defense holds, no caller input is mutated, no nil-map panic path, and the normalize→group-extraction→preprocess ordering is correct. Three findings below: one correctness edge (non-conformant tokens) and two minor cleanups. Comfortable with this PR once finding 1 is addressed or explicitly deemed out of scope.

**Review by JAORMX — APPROVED** [review]
Re-approving on `81090d1` — this revision addresses the feedback from both this pass and @rdimitrov's, and I've verified each:

- **Non-string array elements no longer produce spurious tokens** (@rdimitrov's parity finding): `multiValuedTokens` now keeps only string elements and skips non-strings, instead of `fmt.Sprint`-ing them into `"1e+06"`/`"<nil>"`. Tests updated to match (`non_string_elements_skipped` now expects `"a"` / `["a"]`).
- **`claim_`/`claimset_` extracted to constants** — the collision-freedom invariant is now enforced by shared consts rather than incidental literals.
- **Doc corrected** from "canonical unpadded" to "VERBATIM (spacing/order preserved)" for the string surface, and a clear **Migration hazard** note added warning that listing an array-shaped claim flips `claim_<name>` Set→String and can turn a fail-closed `like` policy into a fail-open substring match — steering operators to `claimset_` + `.contains`. Exactly the concern from the earlier inline.
- The intentional normalize-vs-claimset divergence is now documented so it won't get "deduped" incorrectly.

Real checks are green (Test Go Code, Lint); the only red is the unrelated `E2E Core (core)` group-removal flake (same one seen on #5839 — nothing to do with Cedar authz). Nicely done.

<sub>🤖 AI-assisted panel review via Claude Code (security · OAuth · Go/reuse). Verified against `81090d1`.</sub>
